server/sonar-server/src/main/java/org/sonar/server/computation/component/TreeRootHolderImpl.java

```diff
@@ -19,24 +19,53 @@
  */
 package org.sonar.server.computation.component;
 
+import java.util.HashMap;
+import java.util.Map;
 import java.util.Objects;
 
+import static org.sonar.server.computation.component.DepthTraversalTypeAwareVisitor.Order.POST_ORDER;
+
 /**
  * Holds the reference to the root of the {@link Component} tree for the current CE run.
  */
 public class TreeRootHolderImpl implements MutableTreeRootHolder {
+
   private Component root;
+  private Map<Integer, Component> componentsByRef = new HashMap<>();
 
   @Override
   public void setRoot(Component newRoot) {
     this.root = Objects.requireNonNull(newRoot);
+    feedComponentsByRef();
   }
 
   @Override
   public Component getRoot() {
+    checkRoot();
+    return this.root;
+  }
+
+  public Component getComponentByRef(int ref) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Add the "@Override" annotation above this method signature [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1161)